### ISSUE TITLE
Memory usage spiking when too many concurrent messages are being processed

### ISSUE THREAD

**Issue by molsio** [meltwater/gen_rmq#214]
**Describe the bug**
Publishing 25 msgs to the example consumer crashes the process, and make the process use 100% cpu.

**To Reproduce**
1. Starting this consumer example https://github.com/meltwater/gen_rmq/blob/master/documentation/examples/consumer.ex
2. publish 25 messages,
 0..25
    |> Enum.each(fn e ->
      IO.puts("pub: #{e}")
      Publisher.publish_message(Jason.encode!(%{count: e}), "routing_key.#")
      |> IO.inspect()
      #:timer.sleep(10)
    end)
3 fails with

17:20:18.749 [error] GenServer #PID<0.301.0> terminating
** (stop) :socket_closed_unexpectedly
Last message: :socket_closed
State: {:state, :amqp_network_connection, {:state, #Port<0.6>, "client 127.0.0.1:52430 -> 127.0.0.1:5672", 10, #PID<0.306.0>, 131072, #PID<0.300.0>, :undefined, false}, #PID<0.305.0>, {:amqp_params_network, "guest", {:plaintext, "guest"}, "/", 'localhost', 5672, 2047, 0, 10, 60000, :none, [#Function<9.127867893/3 in :amqp_uri.mechanisms/1>, #Function<9.127867893/3 in :amqp_uri.mechanisms/1>], [], []}, 2047, [{"capabilities", :table, [{"publisher_confirms", :bool, true}, {"exchange_exchange_bindings", :bool, true}, {"basic.nack", :bool, true}, {"consumer_cancel_notify", :bool, true}, {"connection.blocked", :bool, true}, {"consumer_priorities", :bool, true}, {"authentication_failure_close", :bool, true}, {"per_consumer_qos", :bool, true}, {"direct_reply_to", :bool, true}]}, {"cluster_name", :longstr, "rabbit@odin.lan"}, {"copyright", :longstr, "Copyright (c) 2007-2020 VMware, Inc. or its affiliates."}, {"information", :longstr, "Licensed under the MPL 2.0. Website: https://rabbitmq.com"}, {"platform", :longstr, "Erlang/OTP 23.1.1"}, {"product", :longstr, "RabbitMQ"}, {"version", :longstr, "3.8.9"}], :none, false}

4. set concurrency: false solves the problem.
5. set prefetch_count: 1 solves the problem.

**Expected behavior**
Expect the consumer to handle the messages

**Environment**
- Elixir version:  1.10.4 and 1.11
- Erlang/OTP version: 23.1.1
- RabbitMQ version: 3.8.9

**Comment by mkorszun** [comment]
@molsio thanks for reporting. @vorce @akoutmos are you guys able to reproduce it? I am not 😕 I have even created a testcase to do it https://github.com/meltwater/gen_rmq/commit/6e541fff5fdd668cc6369f035925793a6ddf3775 - no luck.

@molsio are you testing against RabbitMQ started via docker-compose we provide?

**Comment by vorce** [comment]
@mkorszun Hmm yes seems I can reproduce it. I use the example consumer and publisher with the docker-compose.yaml from gen_rmq.

**Comment by molsio** [comment]
If it's the docker-compose image in the root of the project, then yes.

Maybe you can reproduce it with my source and guide,

docker run --rm -it --name rabbit -p 5672:5672 -p 5673:5673 -p 15672:15672 rabbitmq:3-management 

git clone git@github.com:molsio/gen_rmq_214_test.git
Cloning into 'gen_rmq_214_test'...
remote: Enumerating objects: 17, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (14/14), done.
Receiving objects: 100% (17/17), 4.78 KiB | 4.78 MiB/s, done.
remote: Total 17 (delta 0), reused 17 (delta 0), pack-reused 0


cd gen_rmq_214_test
mix deps.get

Resolving Hex dependencies...
....

iex -S mix  
Erlang/OTP 23 [erts-11.1.1] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe] [dtrace]
...


iex(2)> Publisher.publish_msgs()
.
.
.
22:02:17.735 [debug] Message successfully processed. Tag: 22
 
22:03:35.488 [debug] [Elixir.Consumer]: Received message. Tag: 24, routing key: routing_key.#, redelivered: false
 
22:03:35.490 [error] GenServer #PID<0.818.0> terminating
** (stop) :socket_closed_unexpectedly
Last message: :socket_closed
State: {:state, :amqp_network_connection, {:state, #Port<0.626>, "client 127.0.0.1:56686 -> 127.0.0.1:5672", 10, #PID<0.825.0>, 131072, #PID<0.817.0>, :undefined, false}, #PID<0.824.0>, {:amqp_params_network, "guest", {:plaintext, "guest"}, "/", 'localhost', 5672, 2047, 0, 10, 60000, :none, [#Function<9.127867893/3 in :amqp_uri.mechanisms/1>, #Function<9.127867893/3 in :amqp_uri.mechanisms/1>], [], []}, 2047, [{"capabilities", :table, [{"publisher_confirms", :bool, true}, {"exchange_exchange_bindings", :bool, true}, {"basic.nack", :bool, true}, {"consumer_cancel_notify", :bool, true}, {"connection.blocked", :bool, true}, {"consumer_priorities", :bool, true}, {"authentication_failure_close", :bool, true}, {"per_consumer_qos", :bool, true}, {"direct_reply_to", :bool, true}]}, {"cluster_name", :longstr, "rabbit@7fd26742b18c"}, {"copyright", :longstr, "Copyright (c) 2007-2020 VMware, Inc. or its affiliates."}, {"information", :longstr, "Licensed under the MPL 2.0. Website: https://rabbitmq.com"}, {"platform", :longstr, "Erlang/OTP 23.1.1"}, {"product", :longstr, "RabbitMQ"}, {"version", :longstr, "3.8.9"}], :none, false}
22:03:35.518 [error] gen_server <0.818.0> terminated with reason: socket_closed_unexpectedly


I'm VERY new to Elixir, so maybe I'm doing something wrong.

**Comment by vorce** [comment]
@mkorszun https://github.com/vorce/gen_rmq214

**Comment by akoutmos** [comment]
Thanks for the repo @vorce. I'll take a look and see what I can make of this.

**Comment by akoutmos** [comment]
I have root caused the issue. It was an interesting one to track down. I'll work on a fix tomorrow and open up a PR.

**Comment by vorce** [comment]
@akoutmos ooh cool! I was planning to dive deeper today. I'm really curious what you found and how :)

**Comment by akoutmos** [comment]
@vorce I laid out what the problem was in PR #215. As to how I tracked it down. I leveraged `Process.info` quite a bit while things were running and noticed the memory consumption growing per message being processed. I also noticed via the telemetry events that the slowdown was occurring when each task was being started (as a result of the growing amount of memory needing to be copied as the state grew recursively).
